### DISCUSSION TITLE
Binding deep nested data doesn't work for two words tables

### DISCUSSION THREAD

**Discussion by malekim** [livewire/livewire#3803] [7. Bugs (archived)]
Shortly, when I try to bind relationship defined in model as camelCase, then Livewire can't handle it.

For example relation for table dogs and dog_diseases:
Table dog:

```
public function dogDiseases()
{
        return $this->hasMany(DogDisease::class, "dog_id", "id");
}
```

dog.php inside Livewire:

```
protected $rules = [
        'dog.dogDiseases.*.name'
];
```

If I change inside table dog name of the relationship function to underscore:

```
public function dog_diseases()
{
        return $this->hasMany(DogDisease::class, "dog_id", "id");
}
```

dog.php inside Livewire:

```
protected $rules = [
        'dog.dog_diseases.*.name'
];
```

It starts working. I should expect that camelCase will work also.

**Comment by joshhanley** [comment]
@malekim can you also share an example of your blade code that has the model binding in it?

Thanks!

**Comment by MattFosterNow** [comment]
Firstly, the deep nested binding is spectacular, so thanks for this! I've managed to get it working on a couple of different components, but like the OP, I'm also experiencing an issue when the nested data comes from a table with a two-word name.

I have two models, Category and FullCategory, representing the categories and full_categories tables in the database. Category has a one-to-many relationship with FullCategory, defined in the Category model using camelCase, like so:

`Category.php`
```
namespace App\Models;

use Illuminate\Database\Eloquent\Model;

class Category extends Model
{
    public function fullCategories()
    {
        return $this->hasMany(FullCategory::class);
    }
}
```
In the Livewire component, the relevant fields for binding are defined in the $rules property, also using camelCase:

`ShowCategory.php`
```
namespace App\Http\Livewire;

use Livewire\Component;
use App\Models\Category;

class ShowCategory extends Component
{
    public $category;
    
    protected $rules = [
        'category.name' => 'required',
        'category.fullCategories.*.name' => 'required',
    ];
...
}
```
And in the Blade view, I'm iterating through the fullCategories, once again using camelCase:

`show-category.blade.php`
```
...
<div>
    <input readonly value="{{ $category->id }}" />
    <input wire:model="category.name" />
</div>
 
@foreach($category->fullCategories as $f => $fullCategory)
    <div wire:key="category.fullCategory-{{ $fullCategory->id }}">
        <input readonly value="{{ $fullCategory->id }}" />
        <input wire:model="category.fullCategories.{{ $f }}.name" />
    </div>
@endforeach
...
```

The category.name field binds correctly, but none of the fullCategory name fields are populated when the component is initialised. However, I noticed that if I start typing in one of the fullCategory name fields, all of a sudden, they all get populated with the correct data. But if I update the category name field, the fullCategories lose their data again.

When I dump the $category property, I can see the relationship is named "full_categories". And when I update a fullCategory name field, the $category property gains a "fullCategories" attribute with all the correct relationship data. So now it has a "fullCategories" attribute as well as a "full_categories" relationship with identical data. But the "fullCategories" attribute disappears again when I update the category name field.

Also, just like the OP mentioned, if I replace all the references to "fullCategories" with "full_categories" in the code, everything works as expected. And in case it's relevant, I have an identical setup (model/component/Blade) that reads from two other tables in the database (both one-word tables) and there are no issues there whatsoever.

I'm using Livewire 2.6.5 and Laravel 8.47. Let me know if I can provide any other info. Thanks!

**Reply by joshhanley** [reply]
@MattFosterNow thanks for all the info! Yeah, it's possible I didn't account for relations with multiple words. I guess under the hood Laravel converts the camel case relation to snake cased when it sets it and we will probably have to handle it the same. I will have a look and see if I can get it running.

Thanks!

**Comment by joshhanley** [comment]
Ok, I've investigated what is going on with this, and as I suspected, Laravel converts relations from camelCased to snake_cased when `toArray()` is called on a model.

So the problem is the rule is in camel case, which matches the relation method and property, but isn't matching up to the data attribute in the array generated from the model. And as Livewire uses those rules to determine what should be sent to the front end, no data is showing up.

Will have to have a think about the best way to handle it so it all works as expected.

**Reply by jeffchown** [reply]
@joshhanley Just came across this scenario in a project of mine.  Has there been any change/update to this item?  Thanks!

**Reply by jeffchown** [reply]
Please disregard my previous post @joshhanley , all good here.

**Reply by joshhanley** [reply]
@jeffchown sorry been away 🙂

Did you manage to sort a work around? If so what did you end up doing?

**Reply by wojciechczerniak** [reply]
My workaround is to overload `toArray` and rename the key:

```php
function toArray() {
    $array = parent::toArray();
    $camelArray = array();
    foreach($array as $name => $value) {
        $camelArray[Str::camel($name)] = $value;
    }
    return $camelArray;
}
```

Although it's probably a matter of time to find that I broke the Laravel somewhere.

Therefore I'm here to ask @joshhanley if you had time/opportunity to look into it? Having references to `shippingAddress` in Laravel and `shipping_address` in Livewire is confusing and error prone. Also, I can't reuse some of the code / validation rules.

**Reply by joshhanley** [reply]
@wojciechczerniak I've tried a few times to get it running, but not been able to get it to work happily 😕

**Reply by luttje** [reply]
Heya, I ran into this problem as well. If it's too challenging to fix, perhaps at least a warning can be built in somehow. It could save others a lot of frustration and time 😭😭

/sarcasm (but only a bit)

**Comment by joshhanley** [comment]
I think I've finally managed to crack this one! Have submitted a PR #5525 and it has been merged. Could someone please test it and let me know how it goes? Thanks!

**Comment by joshhanley** [comment]
We've decided to revert the PR #5525, that fixed this issue, as currently there are too many edge cases. See my comment on PR https://github.com/livewire/livewire/pull/5536#issuecomment-1407813069 for more details on why it was reverted.
